examples/h3client.rs

```diff
@@ -0,0 +1,285 @@
+                debug!("timed out");
+
+                quic_conn.on_timeout();
+
+                break 'read;
+            }
+
+            let len = match socket.recv(&mut buf) {
+                Ok(v) => v,
+
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::WouldBlock {
+                        debug!("recv() would block");
+                        break 'read;
+                    }
+
+                    panic!("recv() failed: {:?}", e);
+                },
+            };
+
+            debug!("{} got {} bytes", quic_conn.trace_id(), len);
+
+            // Process potentially coalesced packets.
+            let read = match quic_conn.recv(&mut buf[..len]) {
+                Ok(v) => v,
+
+                Err(quiche::Error::Done) => {
+                    debug!("{} done reading", quic_conn.trace_id());
+                    break;
+                },
+
+                Err(e) => {
+                    error!("{} recv failed: {:?}", quic_conn.trace_id(), e);
+                    quic_conn.close(false, e.to_wire(), b"fail").unwrap();
+                    break 'read;
+                },
+            };
+
+            debug!("{} processed {} bytes", quic_conn.trace_id(), read);
+        }
+
+        if quic_conn.is_closed() {
+            debug!("{} connection closed", quic_conn.trace_id());
+            break;
+        }
+
+        if quic_conn.is_established() && !req_sent {
+            debug!(
+                "{} QUIC handshake completed, now trying HTTP/3",
+                quic_conn.trace_id()
+            );
+
+            if b"h3-17" != quic_conn.application_proto() {
+                // TODO a better error code?
+                quic_conn
+                    .close(false, 0x0, b"I don't support your ALPNs")
+                    .unwrap();
+                break;
+            }
+
+            // TODO make opening control streams saner
+            if !h3conn.is_established() {
+                h3conn.send_settings(&mut *quic_conn);
+                h3conn.open_qpack_streams(&mut *quic_conn);
+            }
+
+            h3conn.create_placeholder_tree(&mut *quic_conn);
+
+
+            let req = Request::builder()
+                .method("GET")
+                .uri(&uri)
+                .version(http::Version::HTTP_2)
+                .header("User-Agent", "quiche-http/3")
+                .body(())
+                .unwrap();
+
+            info!("Sending HTTP request {:?}", req);
+
+            if let Err(e) = h3conn.send_request(&mut *quic_conn, req, false) {
+                error!("{} stream send failed {:?}", quic_conn.trace_id(), e);
+            } else {
+                req_sent = true;
+            }
+        }
+
+        let streams: Vec<u64> = quic_conn.readable().collect();
+        for s in streams {
+            info!("{} stream id {} is readable", quic_conn.trace_id(), s);
+            let mut h3_frames: Vec<quiche::h3::frame::Frame> = Vec::new();
+            loop {
```

Comment:
I think the per-stream loop should be handled by the library. There's no point in making every application duplicate this loop.

Reply:
I've got two new functions that take the burden of this. Right now they are dumb and read all readables. In future, we should be smarter and selective on what streams to read when, possibly based on QPACK and where the thing is a client or server.